vault/external_tests/raft/raft_autopilot_test.go

```diff
@@ -0,0 +1,212 @@
+package rafttests
+
+import (
+	"context"
+	"math"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/vault/api"
+	"github.com/kr/pretty"
+
+	autopilot "github.com/hashicorp/raft-autopilot"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/hashicorp/vault/helper/namespace"
+	"github.com/hashicorp/vault/helper/testhelpers"
+	"github.com/hashicorp/vault/physical/raft"
+	"github.com/hashicorp/vault/vault"
+)
+
+func TestRaft_Autopilot_Disable(t *testing.T) {
+	cluster := raftCluster(t, &RaftClusterOpts{
+		DisableFollowerJoins: true,
+		InmemCluster:         true,
+		// Not setting EnableAutopilot here.
+	})
+	defer cluster.Cleanup()
+
+	client := cluster.Cores[0].Client
+
+	state, err := client.Sys().RaftAutopilotState()
+	require.NoError(t, err)
+	require.Nil(t, nil, state)
+}
+
+func TestRaft_Autopilot_Stabilization_And_State(t *testing.T) {
+	cluster := raftCluster(t, &RaftClusterOpts{
+		DisableFollowerJoins: true,
+		InmemCluster:         true,
+		EnableAutopilot:      true,
+	})
+	defer cluster.Cleanup()
+
+	// Check that autopilot execution state is running
+	client := cluster.Cores[0].Client
+	state, err := client.Sys().RaftAutopilotState()
+	require.NoError(t, err)
+	require.Equal(t, api.AutopilotRunning, state.ExecutionStatus)
+	require.Equal(t, true, state.Healthy)
+	require.Len(t, state.Servers, 1)
+	require.Equal(t, "core-0", state.Servers["core-0"].ID)
+	require.Equal(t, "alive", state.Servers["core-0"].NodeStatus)
+	require.Equal(t, "leader", state.Servers["core-0"].Status)
+
+	config, err := client.Sys().RaftAutopilotConfiguration()
+	require.NoError(t, err)
+
+	// Wait for 110% of the stabilization time to add nodes
+	stabilizationKickOffWaitDuration := time.Duration(math.Ceil(1.1 * float64(config.ServerStabilizationTime)))
+	time.Sleep(stabilizationKickOffWaitDuration)
+
+	joinAndStabilizeFunc := func(core *vault.TestClusterCore, nodeID string, numServers int) {
+		joinFunc := func(core *vault.TestClusterCore) {
+			_, err := core.JoinRaftCluster(namespace.RootContext(context.Background()), []*raft.LeaderJoinInfo{
+				{
+					LeaderAPIAddr: client.Address(),
+					TLSConfig:     cluster.Cores[0].TLSConfig,
+					Retry:         true,
+				},
+			}, false)
+			require.NoError(t, err)
+			time.Sleep(1 * time.Second)
+			cluster.UnsealCore(t, core)
+		}
+		joinFunc(core)
+
+		state, err = client.Sys().RaftAutopilotState()
+		require.NoError(t, err)
+		require.Equal(t, false, state.Healthy)
+		require.Len(t, state.Servers, numServers)
+		require.Equal(t, false, state.Servers[nodeID].Healthy)
+		require.Equal(t, "alive", state.Servers[nodeID].NodeStatus)
+		require.Equal(t, "non-voter", state.Servers[nodeID].Status)
+
+		// Wait till the stabilization period is over
```

Comment:
Should we be testing to make sure the server doesn't become a voter before the configured stabilization time?